### ISSUE TITLE
Cannot compile foma back-end

### ISSUE THREAD

**Issue by reynoldsnlp** [reynoldsnlp/hfst-python#4]
No matter whether I include foma's `C` or `C++` variant, compilation fails. The choice between variants is in `setup.py`.

When I include the C version, the error is undeclared `disclaimer1`, `SVN_REV`, and `disclaimer2`, all on the same line.

When I include the C++ version, the errors mainly surround incompatibilities of `void *` (see below).

`C` errors:

```
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_FOMA -DHAVE_OPENFST -DHAVE_OPENFST_LOG -DHAVE_READLINE -DHAVE_GETLINE -I/tmp/pip-req-build-ifj1cd37/src/hfst -I/tmp/pip-req-build-ifj1cd37/src/hfst/lib -I/tmp/pip-req-build-ifj1cd37/libhfst_src/back-ends/foma -I/tmp/pip-req-build-ifj1cd37/libhfst_src/back-ends -I/tmp/pip-req-build-ifj1cd37/libhfst_src/libhfst/src -I/tmp/pip-req-build-ifj1cd37/libhfst_src/libhfst/src/parsers -I/tmp/pip-req-build-ifj1cd37/libhfst_src/back-ends/openfst/src/include -I/opt/_internal/cpython-3.6.13/include/python3.6m -c libhfst_src/back-ends/foma/foma.c -o build/temp.linux-x86_64-3.6/libhfst_src/back-ends/foma/foma.o -Wno-sign-compare
    libhfst_src/back-ends/foma/foma.c: In function ‘main’:
    libhfst_src/back-ends/foma/foma.c:166:41: error: ‘disclaimer1’ undeclared (first use in this function); did you mean ‘disclaimer’?
      166 |       printf("%s%i.%i.%i%s (svn r%i)%s",disclaimer1,MAJOR_VERSION,MINOR_VERSION,BUILD_VERSION,STATUS_VERSION,SVN_REV,disclaimer2);
          |                                         ^~~~~~~~~~~
          |                                         disclaimer
    libhfst_src/back-ends/foma/foma.c:166:41: note: each undeclared identifier is reported only once for each function it appears in
    libhfst_src/back-ends/foma/foma.c:166:110: error: ‘SVN_REV’ undeclared (first use in this function)
      166 |       printf("%s%i.%i.%i%s (svn r%i)%s",disclaimer1,MAJOR_VERSION,MINOR_VERSION,BUILD_VERSION,STATUS_VERSION,SVN_REV,disclaimer2);
          |                                                                                                              ^~~~~~~
    libhfst_src/back-ends/foma/foma.c:166:118: error: ‘disclaimer2’ undeclared (first use in this function); did you mean ‘disclaimer’?
      166 |       printf("%s%i.%i.%i%s (svn r%i)%s",disclaimer1,MAJOR_VERSION,MINOR_VERSION,BUILD_VERSION,STATUS_VERSION,SVN_REV,disclaimer2);
          |                                                                                                                      ^~~~~~~~~~~
          |                                                                                                                      disclaimer
```

`C++` errors:

```
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:154:12: error: assigning to 'struct lexc_hashtable *' from incompatible type 'void *'
        hnew = xxmalloc(sizeof(struct lexc_hashtable));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:172:17: error: assigning to 'struct lexc_hashtable *' from incompatible type 'void *'
        hashtable = xxcalloc(SIGMA_HASH_TABLESIZE, sizeof(struct lexc_hashtable));
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:176:14: error: assigning to 'bool *' from incompatible type 'void *'
        mchash = xxcalloc(256*256, sizeof(Boolean));
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:192:10: error: assigning to 'struct statelist *' from incompatible type 'void *'
        sl = xxmalloc(sizeof(struct statelist));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:217:21: error: assigning to 'struct states::trans *' from incompatible type 'void *'
                    newtrans = xxmalloc(sizeof(struct states::trans));
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:246:18: error: assigning to 'int *' from incompatible type 'void *'
        sigreplace = xxcalloc(sigma_max(net->sigma)+1,sizeof(int));
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:272:15: error: assigning to 'int *' from incompatible type 'void *'
            unk = xxcalloc(sigma_max(lexsigma)+2,sizeof(int));
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:281:13: error: assigning to 'struct states **' from incompatible type 'void *'
        slist = xxcalloc(sizeof(**slist),maxstate+1);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:282:14: error: assigning to 'int *' from incompatible type 'void *'
        finals = xxcalloc(sizeof(int),maxstate+1);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:285:20: error: assigning to 'struct states *' from incompatible type 'void *'
            newstate = xxmalloc(sizeof(struct states));
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:294:13: error: assigning to 'struct statelist *' from incompatible type 'void *'
            s = xxmalloc(sizeof(struct statelist));
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:302:16: error: assigning to 'struct states::trans *' from incompatible type 'void *'
        newtrans = xxmalloc(sizeof(struct states::trans));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:312:24: error: assigning to 'struct states::trans *' from incompatible type 'void *'
                newtrans = xxmalloc(sizeof(struct states::trans));
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:323:36: error: assigning to 'struct states::trans *' from incompatible type 'void *'
                            newtrans = xxmalloc(sizeof(struct states::trans));
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:338:24: error: assigning to 'struct states::trans *' from incompatible type 'void *'
                newtrans = xxmalloc(sizeof(struct states::trans));
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:378:9: error: assigning to 'struct lexstates *' from incompatible type 'void *'
        l = xxmalloc(sizeof(struct lexstates));
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:384:16: error: assigning to 'struct states *' from incompatible type 'void *'
        newstate = xxmalloc(sizeof(struct states));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:679:17: error: assigning to 'struct multichar_symbols *' from incompatible type 'void *'
            mcnew = xxmalloc(sizeof(struct multichar_symbols));
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    libhfst_src/back-ends/foma/cpp-version/lexcread.cc:752:20: error: assigning to 'struct states::trans *' from incompatible type 'void *'
            newtrans = xxmalloc(sizeof(struct states::trans));
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    4 warnings and 20 errors generated.
```

**Comment by reynoldsnlp** [comment]
It looks like `libhfst_src/back-ends/foma/foma.c` is missing the following lines that are present in `libhfst_src/back-ends/foma/cpp-version/foma.cc`:

```c++
char disclaimer1[] = "Foma, version ";
char disclaimer2[] = "\nCopyright © 2008-2015 Mans Hulden\nThis is free software; see the source code for copying conditions.\nThere is ABSOLUTELY NO WARRANTY; for details, type \"help license\"\n\nType \"help\" to list all commands available.\nType \"help <topic>\" or help \"<operator>\" for further help.\n\n";

#ifndef SVN_REV
#define SVN_REV 0
#endif
```

**Comment by reynoldsnlp** [comment]
Looks like there are efforts afoot to decouple foma anyway: https://github.com/hfst/hfst/issues/485
@TinoDidriksen is suggesting using a Debian package there. I'm not sure what he has in mind for MacOS and Windows.

**Comment by TinoDidriksen** [comment]
foma.c is not needed for the library. Only the files listed in https://github.com/hfst/hfst/blob/master/back-ends/foma/Makefile.am#L30 are.

I guess nobody builds HFST with `GENERATE_FOMA_WRAPPER`, or this should have errored out.

**Comment by reynoldsnlp** [comment]
Ah, I should have been more careful about checking which files were listed in the old setup.py instead of just globbing everything.
